src/current/v23.2/plpgsql.md

```diff
@@ -0,0 +1,446 @@
+{{site.data.alerts.end}}
+
+- `level` is the message severity. Possible values are `DEBUG`, `LOG`, `NOTICE`, `INFO`, `WARNING`, and `EXCEPTION`. Specify `EXCEPTION` to raise an error that aborts the current transaction.
+- `message` is a message string to display.
+- `expressions` is an optional, comma-separated list of [expressions](https://www.postgresql.org/docs/16/plpgsql-expressions.html) that provide values to replace any `%` placed within the message string. The number of expressions must match the number of `%` placeholders.
+- `option` is a type of additional information to include. Possible values are `MESSAGE`, `DETAIL`, `HINT`, or `ERRCODE`. To specify `MESSAGE`, use the following alternate syntax: 
+
+	~~~ sql
+	RAISE level USING MESSAGE = 'message';
+	~~~
+
+- `expression` is an expression to display that corresponds to the specified `option`. If `ERRCODE` is the specified option, this must be a valid [`SQLSTATE` error code or name](https://www.postgresql.org/docs/16/errcodes-appendix.html).
+
+For example:
+
+{% include_cached copy-clipboard.html %}
+~~~ sql
+CREATE OR REPLACE PROCEDURE raise_time() AS $$
+  BEGIN
+    RAISE NOTICE 'current timestamp: %', now()
+	USING HINT = 'Call this procedure again for a different result';
+  END
+  $$ LANGUAGE PLpgSQL;
+~~~
+
+{% include_cached copy-clipboard.html %}
+~~~
+CALL raise_time();
+~~~
+
+~~~
+NOTICE: current timestamp: 2024-01-05 23:09:08.0601+00
+HINT: Call this procedure again for a different result
+CALL
+~~~
+
+#### Write exception logic
+
+Use an `EXCEPTION` statement to catch and handle specified errors. 
+
+Any valid [`SQLSTATE` error code or name](https://www.postgresql.org/docs/16/errcodes-appendix.html) can be specified, except for Class 40 (transaction rollback) errors. Arbitrary user-defined `SQLSTATE` codes can also be specified.
+
+If a specified error is caught, the exception handling statements are executed. Any unspecified errors are caught by `WHEN OTHERS`, except for `query_canceled` and `assert_failure`.
+
+~~~ sql
+EXCEPTION
+	WHEN error THEN
+		handle_exception;
+	[ WHEN error_n THEN
+		handle_exception_n; ]
+	[ WHEN OTHERS THEN
+		handle_other_exceptions; ]
+~~~
+
+`EXCEPTION` logic is typically included after the main body of a PL/pgSQL block. For example:
+
+~~~ sql
+BEGIN
+	...
+  EXCEPTION
+    WHEN not_null_violation THEN
+      RETURN 'not_null_violation';
+    WHEN OTHERS THEN
+      RETURN others;
+  END
+~~~
+
+## Examples
+
+### Create a user-defined function using PL/pgSQL
+
+{% include {{ page.version.version }}/sql/udf-plpgsql-example.md %}
+
+### Create a stored procedure using PL/pgSQL
+
+{% include {{page.version.version}}/sql/movr-statements.md %}
+
+{% include {{ page.version.version }}/sql/stored-proc-example.md %}
+
+## Known limitations
+
+- PL/pgSQL blocks cannot be nested.
+- Cursors used in PL/pgSQL always materialize their contents eagerly on opening. This can affect performance and resource usage.
```

Comment:
ignorable nit: not sure what we are trying to say here: "materialize their contents eagerly".

Reply:
The result of a PL/pgSQL cursor is fully evaluated up front and stored in a container. For SQL cursors, the evaluation is lazy, and results are pulled only as they are requested via FETCH or MOVE.

Reply:
Amending to "Cursors used in PL/pgSQL execute their queries on opening".